cityofaustin/atd-moped
# Fixes team member role editing

## Associated issues
https://github.com/cityofaustin/atd-data-tech/issues/10474

This PR adds a new table, `moped_proj_personnel_roles` which enables a normalized many-to-many relationship between project team members (`moped_proj_personnel`) and project roles (`moped_project_roles`).

Prior to this change, we actually stored duplicate personnel records, one per project role, and concatenated their details into a single table row in the editor. This PR includes a migration script that de-dupes the personnel records and creates their `moped_proj_personnel_roles`.

Once this has been merged, we'll need to create a separate migration that removes the depreacted `role_id` column from `moped_proj_personnel`.

A few other UI tweaks:
- show role descriptions in menu options
- autofocus team member input after adding a row
- prevent same team member from being added to the table twice

## Testing
**URL to test:**  Local



**Steps to test the editor UI:**
1. Add a team member to the project team with an **Inspector** role.
2. Verify their workgroup populates once the row is saved.
3. Edit the team member to remove their **Insepctor** role and add **Designer** and **Project Owner** roles. Verify the changes save correctly
4. Add another team member—make them an **Inspector**.
5. Take note of the names of your Inspector and Designer. Navigate to the Projects list, and verify these names populate in their respective columns on the team table. 
6.  Use the advanced search to search for your project based on team members you've added to your project.
7. Delete one of the team members from your project.
8. Verify the deleted team member is removed from the Team tab and the Projects list
9. Stop the Hasura cluster

**Steps to test the migration script:**
1. Start the hasura cluster
2. Activate a python env with `requests` installed
3. Navigate to `/moped-toolbox/migrate_personnel`
4. `python run.py -e local`
5. In the editor, check a few projects and observe that the team members look normal.
6. Edit these existing team members and confirm that they work normally.


Pre-merge todos:
- [ ] Verify that `not null` constraint migrations will succeed in staging and prod
- [ ] Update seed data

---
#### Ship list
